### PR TITLE
Add max_supply to respective minters

### PR DESCRIPTION
**Pull request by tanzhuoyao-nus**

### CODE REVIEW

File: src/tbm/CommunityMinter.scilla [hunk 1]
```diff
@@ -176,6 +182,19 @@ procedure IsWithinMintLimit(quantity: Uint32)
   end
 end
 
+procedure IsNotExceedingMaxCommunitySupply(quantity: Uint32)
```
Comment (RavenXce) [inline src/tbm/CommunityMinter.scilla:185]: should be called `SetAndValidateTotalSupply` as this mutates state

File: src/tbm/CommunityMinter.scilla [hunk 2]
```diff
@@ -176,6 +182,19 @@ procedure IsWithinMintLimit(quantity: Uint32)
...
+  current_supply <- total_community_supply;
+  new_supply = builtin add current_supply quantity;
+  is_not_exceeding = uint32_le new_supply max_community_supply;
```
Comment (RavenXce) [inline src/tbm/CommunityMinter.scilla:188]: this means the max mint count is actually one less than `max_community_supply`?

File: src/tbm/GiveawayMinter.scilla [hunk 3]
```diff
@@ -55,6 +55,7 @@ type Error =
   | CodeNotPendingOwner
   | CodePendingOwnerNotEmpty
   | CodeExceedMaxMintQuantity
+  | CodeMaxGiveawaySupplyExceeded
```
Comment (RavenXce) [inline src/tbm/GiveawayMinter.scilla:58]: Same as above

**Review by RavenXce — CHANGES_REQUESTED** [review]

File: src/tbm/NFT.scilla [hunk 4]
```diff
@@ -133,9 +163,18 @@ field token_uris: Map Uint256 String = Emp Uint256 String
...
+field total_reserved: Uint256 = Uint256 0
+
 (* Base URI that will be updated once upon reveal *)
 field base_uri: String = "https://d5gozuhfenzxt.cloudfront.net/metadata/"
 
+(* Offset index after sale ends *)
+field offset_index: Uint256 = Uint256 0
+
+(* One-time flag to ensure offset applied only once *)
+field is_offset_applied: Bool = False
```
Comment (stevenkhong) [inline src/tbm/NFT.scilla:176]: not necessary, use `offset_index > 0` as check for whether value is set.

File: src/tbm/GiveawayMinter.scilla [hunk 5]
```diff
@@ -147,23 +150,46 @@ procedure IsWithinMintLimit(quantity: Uint32)
   end
 end
 
+procedure IsReservedToken(token_id: Uint256)
```
Comment (stevenkhong) [inline src/tbm/GiveawayMinter.scilla:153]: is this check necessary? imo use a `distributed_tokens: Uint256`  will do, increment at `DistributeToken` rather then expecting `token_id`. and check for `distributed_tokens` < `reserved_supply` instead

File: src/tbm/NFT.scilla (hunk 4 above)
Comment (RavenXce) [reply src/tbm/NFT.scilla:176]: @tanzhuoyao-nus use `Option {Uint256}` if need be

File: src/tbm/NFT.scilla [hunk 6]
```diff
@@ -133,6 +170,12 @@ field token_uris: Map Uint256 String = Emp Uint256 String
 (* Total supply of tokens *)
 field total_supply: Uint256 = Uint256 0
 
+(* Tokens reserved to be minted by GiveawayMinter *)
```
Comment (RavenXce) [inline src/tbm/NFT.scilla:173]: Can do `field public_supply: Uint256 = builtin sub max_supply reserved_supply`

**Pull request merged**